pytorch_lightning/trainer/connectors/signal_connector.py

```diff
@@ -1,14 +1,30 @@
 import logging
 import os
 import signal
+import sys
 from subprocess import call
 
+from pytorch_lightning.plugins.environments import SLURMEnvironment
+from pytorch_lightning.utilities.imports import _fault_tolerant_training
+
 log = logging.getLogger(__name__)
 
 
-class SLURMConnector:
+class SignalConnector:
     def __init__(self, trainer):
         self.trainer = trainer
+        self.trainer._should_gracefully_terminate = False
+
+    def register_signal_handlers(self):
+        cluster_env = getattr(self.trainer.training_type_plugin, "cluster_environment", None)
+        if isinstance(cluster_env, SLURMEnvironment):
+            self.register_slurm_signal_handlers()
+        elif _fault_tolerant_training() and not sys.platform == "win32":
+            self.register_fault_tolerant_handlers()
+
+    def register_fault_tolerant_handlers(self):
+        signal.signal(signal.SIGUSR1, self.sig_fault_tolerant_handler)
+        signal.signal(signal.SIGTERM, self.term_handler)
```

Comment:
It's very typical for the user script or the runtime environment to register these signal handlers. It may lead to issues if a library overrides already registered custom signal handlers without the user knowing. For example, one of FB's production environments relies on SIGTERM handler to collect diagnostic information about the job, which this change will break.

Some alternatives I can think of:
- Instead of having the trainer register the signal handlers automatically, offer a utility for the user to register it themselves. This gives the user a chance to merge/reconcile multiple signal handlers.
- Save the previously registered signal handlers and invoke them when the trainer is ready to terminate gracefully. A catch though is that invoking a signal handler at a different place/time may not always make sense.

Reply:
Good point. I added a mechanism to bypass it.